### PR TITLE
Refactor simulation tools configuration

### LINKED ISSUE #792: simtools-simulate-prod does not accept required parameters defined through config file
Perhaps an issue, but perhaps this is intentional:

- [ ] **Issue 1**:
` --primary, --azimuth_angle, --zenith_angle` are required for running `simtools-simulate-prod`. However, these arguments may also be defined in the production config file instead (`--production_config prod_config_proton_mini.yml` as in my case). I think the code should recognize that these three required parameters were already passed in the config file.
- [ ] **Issue 2**:
running without passing --site leads to an error, even though site is also defined in the config file
- [x] **Issue 3**: (solved, see comment below)
already for nshow=10 it seems the output it empty (from the config file) and nshow=100 from the command line_

### LINKED ISSUE #911: Simulator configurator unecessary complicated and requires cross-module knowledge of data structures
The configuration of the Simulator tools through a yml file (e.g., ./tests/resources/prod_multi_config_test_North.yml) is unnecessary complicated. 

The config dictionaries are modified in the applications (e.g., production.py or simulate_prod.py) and in Simulator in order to generate the right format for CorsikaRunner and SimtelRunner. High-level tools / modules need therefore to know exactly the data structure of the low level tools. 

At several levels, there is a large number of modifications to the configuration dicts, including fine-tuned additions or pops. Very hard to maintain and we need to simplify this.

### LINKED ISSUE #982: Refactor simulation configuration
The following outlines a large refactoring project for the simulation configuration code. Any input would be appreciated.

The current implementation of the configuration of simulation productions / runs needs updates to take additional functionality into account since its first implementation:

- improved command line configurator
- usage of json-schema for yaml file validation (e.g., data/parameters/simtel-runner-array_parameters.yml and data/parameters/array-simulator_parameters is used as template for validation only)
- the possibility to update model parameters using the model parameter repository (meaning functionality of the "array block" to change parameter (see e.g., [here](https://github.com/gammasim/simtools/blob/43ae3339fcb6b9cf557fd0a02417b691778780cc/tests/resources/array_config_test.yml#L17) is not required anymore)
- corsika and sim_telarray configuration parameters will go / are already in the database (e.g., [here](https://gitlab.cta-observatory.org/cta-science/simulations/simulation-model/model_parameters/-/tree/main/model_versions/2024-02-01/verified_model/configuration_sim_telarray/LSTN-design?ref_type=heads))

Given this, I will try and implement the following changes to address the points discussed above and the issues in [this milestone](https://github.com/gammasim/simtools/milestone/14):

- move all configuration parameters into the database (add a `configuration_corsika` section); means also to add schema for all of them (these are e.g., many of values in `INTERACTION_FLAGS`, `CHERENKOV_EMISSION_PARAMETERS`, `DEBUGGING_OUTPUT_PARAMETERS` sections of [data/parameters/corsika_parameters.yml](data/parameters/corsika_parameters.yml))
- use only the command line configurator for the configuration of the simulation productions, and not any other yaml files (command line configuration parameters can be provided using a config file)

After implementing these changes, the following example configuration files will be obsolete:

- tests/resources/array_config_test.yml
- data/parameters/corsika_parameters.yml
- tests/resources/corsikaConfigTest.yml
- tests/resources/corsikaConfigTest_astropy_headers.yml
- tests/resources/prod_multi_config_test_North.yml
- tests/resources/prod_multi_config_test_South.yml
- data/parameters/simtel-runner-array_parameters.yml
- data/parameters/array-simulator_parameters

This should allow to remove some of the code dealing with configuration dictionaries and their validation (e.g., general.validate_config_data)

### PR DESCRIPTION
**Pull request by GernotMaier**
Simplifies the configuration of simulation tools. 

This is a major refactoring PR touching almost all modules relevant for simulation configuration and production. This addresses especially `simulate_prod`.

Main changes are:

- removal of configuration by yaml file (as those listed in data/parameters); replaced by configuration via command line (or using configuration files for all command line parameters). There are two parameter files left (camera-efficiency_parameters.yml, corsika_parameters.yml), which will be removed in upcoming PRs
- re-organiation of all 'runners': there is now a dedicated runner directory to collect all of them
- introduce module `runner_services` to collect common functionality
- removed a lot of obsolete code, unused code; simplified many models (e.g., instead replaced get_output_file, get_log_file, get_histogram_file, etc by get_file(file_type="output").
- removed application `simulate_showers_for_trigger_rates` - this is a duplication of `simulate_prod`.
- removed a couple of debug statements which were part of loops (to many output statements makes it hard to read)
- trying to replace module calls like `_check_run_result(self, **kwargs)` by `def _check_run_result(self, run_number=None):` - this makes it much clearer
- rigorous testing of all new functions / modules

Closes #982
Closes #792 (configuration is now completely different and this does not apply anymore)
Closes #911 (configuration files are now handled through the command line and argparser; no metaschema required)

### CODE REVIEW

File: simtools/applications/simulate_prod.py [hunk 1]
```diff
@@ -120,80 +101,6 @@ def _parse(description=None):
...
-        type=int,
-        required=True,
-    )
-    config.parser.add_argument(
-        "--run",
-        help="Run number (actual run number will be 'start_run' + 'run')",
-        type=int,
-        required=True,
-    )
     config.parser.add_argument(
```
Comment (tobiaskleiner) [inline simtools/applications/simulate_prod.py:104]: All the removed arguments still appear in the docstring above. Why did you remove the `config.parser.add_argument`s?

File: simtools/applications/simulate_prod.py [hunk 2]
```diff
@@ -221,72 +168,25 @@ def main():  # noqa: D103
...
-    config_data["common"]["phi"] = args_dict["azimuth_angle"]
-    label = config_data["common"].pop("label", "test-production")
-    config_data["common"]["data_directory"] = Path(args_dict["data_directory"]) / label
-
-    if args_dict["nshow"] is not None:
-        config_data["showers"]["nshow"] = args_dict["nshow"]
-    if args_dict["label"] is not None:
-        label = args_dict["label"]
-
     simulator = Simulator(
```
Comment (tobiaskleiner) [inline simtools/applications/simulate_prod.py:171]: The above comment is related to this: We don't want to keep the option to set the parameters by hand and overwrite the default settings?

File: simtools/applications/simulate_showers_for_trigger_rates.py [hunk 3]
```diff
@@ -1,232 +0,0 @@
-#!/usr/bin/python3
-r"""
```
Comment (tobiaskleiner) [inline simtools/applications/simulate_showers_for_trigger_rates.py:2]: Completely removed this one?

Comment (GernotMaier) [reply simtools/applications/simulate_showers_for_trigger_rates.py:2]: Yes - it simulates corsika showers. This is something simulate_prod can do - the functionality was duplicated.

File: simtools/corsika/corsika_config.py [hunk 4]
```diff
@@ -697,3 +433,49 @@ def get_corsika_input_list(self):
...
+            Run number.
+
+        Raises
+        ------
+        ValueError
+            If run_number is not a valid value (e.g., < 1).
+        """
+        if run_number is None:
+            return self.run_number
+        if not float(run_number).is_integer() or run_number < 1 or run_number > 999999:
```
Comment (tobiaskleiner) [inline simtools/corsika/corsika_config.py:477]: Can we have several runs simultaneausly if we use divergent pointing?

File: simtools/simulator.py [hunk 5]
```diff
@@ -713,27 +522,5 @@ def _get_runs_to_simulate(self, run_list=None, run_range=None):
 
         """
         if run_list is None and run_range is None:
-            if self.runs is None:
-                msg = "Runs to simulate were not given as arguments nor in config_data - aborting"
```
Comment (tobiaskleiner) [inline simtools/simulator.py:717]: no error message?

File: tests/unit_tests/test_simulator.py [hunk 6]
```diff
@@ -35,117 +30,102 @@ def corsika_file():
...
+    args_dict["array_layout_name"] = "test_layout"
+    args_dict["site"] = "North"
+    args_dict["keep_seeds"] = False
+    args_dict["run_number_start"] = 1
+    args_dict["nshow"] = 10
+    return args_dict
+
+
+@pytest.fixture()
+def array_simulator(io_handler, db_config, simulations_args_dict):
```
Comment (tobiaskleiner) [inline tests/unit_tests/test_simulator.py:49]: simulator_array?

**Review by tobiaskleiner — COMMENTED** [review]
Thanks @GernotMaier , see few comments.

File: simtools/simulator.py (hunk 5 above)
Comment (GernotMaier) [reply simtools/simulator.py:717]: Raise now a 'ValueError' for a zero length run list.

File: simtools/corsika/corsika_config.py (hunk 4 above)
Comment (GernotMaier) [reply simtools/corsika/corsika_config.py:477]: I am not sure I understand your comment. Can you explain a little bit what you mean? Why is this connected to divergent pointing?

File: simtools/camera_efficiency.py [hunk 7]
```diff
@@ -155,15 +155,8 @@ def _load_files(self):
...
             self._file[label] = self._base_directory.joinpath(file_name)
 
-    def simulate(self, force=False):
-        """
-        Simulate camera efficiency using testeff.
-
-        Parameters
-        ----------
-        force: bool
-            Force flag will remove existing files and simulate again.
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/camera_efficiency.py:165]: Simulations take sometimes very long. You don't want to remove a 30 hours simulation run by accident, this is why we want to be careful.

File: simtools/applications/simulate_prod.py (hunk 2 above)
Comment (GernotMaier) [reply simtools/applications/simulate_prod.py:171]: Same here, this functionality is simply moved to the command_line module (and therefore generalized to avoid duplication).

File: simtools/applications/simulate_prod.py (hunk 1 above)
Comment (GernotMaier) [reply simtools/applications/simulate_prod.py:104]: No, they are moved to the command_line module, as we use this type of parameters for several applications. So the docstring is still fine (as e.g., you want to configure simulate_prod with azimuth or number of runs).

**Comment by GernotMaier** [comment]
@tobiaskleiner - thanks a lot for the review. I am waiting for one clarification from your side (on the run numbers / divergent pointing issue). Let me know if there is anything else.

File: simtools/corsika/corsika_config.py (hunk 4 above)
Comment (tobiaskleiner) [reply simtools/corsika/corsika_config.py:477]: Maybe having >200 runs per night (if we define several runs parallel), can lead to exceeding this limit over the lifetime of CTAO. Not sure this is realistic though.

File: simtools/camera_efficiency.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tobiaskleiner) [reply simtools/camera_efficiency.py:165]: Ok thans, makes sense.

File: simtools/corsika/corsika_config.py (hunk 4 above)
Comment (GernotMaier) [reply simtools/corsika/corsika_config.py:477]: This is a corsika limit! Nothing we can do about, corsika does not allow larger run numbers (Note that there is no meaning for run numbers, we can simulate all runs with the same run number and things would be ok).

File: simtools/applications/simulate_prod.py (hunk 1 above)
Comment (tobiaskleiner) [reply simtools/applications/simulate_prod.py:99]: In the docstring the parameters are under command_line_parameters. Maybe you can make clear that the values come from the config file.

File: simtools/corsika/corsika_config.py (hunk 4 above)
Comment (tobiaskleiner) [reply simtools/corsika/corsika_config.py:477]: Ah ok I see. Then all good here.

File: simtools/applications/simulate_prod.py (hunk 1 above)
Comment (GernotMaier) [reply simtools/applications/simulate_prod.py:99]: This is correct, they are command line parameters:

- any argument defined through Configurator can be set on the command line, by environmental variable or using the `--config` option and given them all in a configuration file.

This allows us to have let the user decide to configure tools with super long command lines or using a configuration tool.

Comment (tobiaskleiner) [reply simtools/applications/simulate_prod.py:99]: I would say they are converted into command line parameters if they are passed via a configuration file? Maybe you can add this sentence at the beginning somewhere.

Comment (GernotMaier) [reply simtools/applications/simulate_prod.py:95]: This has been changed, as discussed in the chat.

**Comment by GernotMaier** [comment]
Thanks again @tobiaskleiner  . Let me know if there is anything else.

**Review by tobiaskleiner — APPROVED** [review]
All good now. Thanks again @GernotMaier.
